### PR TITLE
(maint) Prevent `generate` and `import` from overwriting existing CA files

### PR DESCRIPTION
**Pull request by Magisus**
Supersedes https://github.com/puppetlabs/puppetserver-ca-cli/pull/18

### CODE REVIEW

File: lib/puppetserver/ca/action/generate.rb [hunk 1]
```diff
@@ -143,15 +140,25 @@ def generate_pki(settings, signing_digest, subject_alt_names = '')
             [settings[:cakey], int_key],
           ]
 
+          errors = FileSystem.check_for_existing_files(public_files.map { |f| f.first })
```
Comment (justinstoller) [inline lib/puppetserver/ca/action/generate.rb]: little golf-y but `(&:first)` works too.

File: lib/puppetserver/ca/action/generate.rb [hunk 2]
```diff
@@ -143,15 +140,25 @@ def generate_pki(settings, signing_digest, subject_alt_names = '')
...
+          errors += FileSystem.check_for_existing_files(private_files.map { |f| f.first })
+
+          if errors.any?
```
Comment (justinstoller) [inline lib/puppetserver/ca/action/generate.rb]: Definitely more readable but I'm still don't love using `any?` like this

File: lib/puppetserver/ca/action/import.rb [hunk 3]
```diff
@@ -65,40 +79,29 @@ def run(input)
...
+          results = {}
+          parser = self.class.parser(results)
 
-        errors = CliParsing.parse_with_errors(parser, args)
+          errors = CliParsing.parse_with_errors(parser, args)
 
-        if check_flag_usage(results)
-          errors << check_flag_usage(results)
-        end
+          if check_flag_usage(results)
```
Comment (justinstoller) [inline lib/puppetserver/ca/action/import.rb]: I think elsewhere we updated this to be
```rb
if error = check_flag_usage(results)
  errrors << error
end
```
if we did that we probably want to do the same here for consistency, though I could see us preferring
```rb
error = check_flag_usage(results)
errors << error if error
```

File: lib/puppetserver/ca/action/generate.rb [hunk 4]
```diff
@@ -123,18 +122,16 @@ def generate_pki(settings, signing_digest, subject_alt_names = '')
...
             [settings[:cacrl], [int_crl, root_crl]],
             [settings[:hostcert], master_cert],
             [settings[:localcacert], [int_cert, root_cert]],
             [settings[:localcacrl], [int_crl, root_crl]],
-            [settings[:hostpubkey], [master_key.public_key]],
-            [settings[:capub], [int_key.public_key]],
+            [settings[:hostpubkey], master_key.public_key],
+            [settings[:capub], int_key.public_key],
+            [settings[:cert_inventory], inventory_entry(master_cert)],
+            [settings[:serial], "002"],
```
Comment (justinstoller) [inline lib/puppetserver/ca/action/generate.rb]: nit: looking at the code we actually do the inventory serial number format here as well in the puppetserver code. We might want to do that here or as a separate ticket (kind of an unrelated bug fix).

File: lib/puppetserver/ca/action/generate.rb [hunk 5]
```diff
@@ -143,15 +140,25 @@ def generate_pki(settings, signing_digest, subject_alt_names = '')
             [settings[:cakey], int_key],
           ]
 
+          errors = FileSystem.check_for_existing_files(public_files.map { |f| f.first })
+          errors += FileSystem.check_for_existing_files(private_files.map { |f| f.first })
+
+          if errors.any?
+            errors << "If you would really like to replace your CA, please delete the existing files first.
```
Comment (justinstoller) [inline lib/puppetserver/ca/action/generate.rb]: This is the preferred way to do it in clojure, but for some reason I thought HEREDOCs were the preferred way to do multiline strings with newlines in them in Ruby. I can't find any consensus on the Internet or even recommendations in Rubocop. So I'm probably wayy off on that. Do you know if there's any preference within the company?

File: lib/puppetserver/ca/action/generate.rb (hunk 1 above)
Comment (Magisus) [reply lib/puppetserver/ca/action/generate.rb]: Oh yeah, always forget about that.

File: lib/puppetserver/ca/action/generate.rb (hunk 2 above)
Comment (Magisus) [reply lib/puppetserver/ca/action/generate.rb]: I missed the conversation on why that is last time. What don't you like about it? (I'm happy to change it, just would like to know what the misgivings are.)

File: lib/puppetserver/ca/action/import.rb (hunk 3 above)
Comment (Magisus) [reply lib/puppetserver/ca/action/import.rb]: I like the explicit block better, I kind of don't like the post-condition syntax most of the time. I'll update this to be like it is elsewhere.

File: lib/puppetserver/ca/action/generate.rb (hunk 4 above)
Comment (Magisus) [reply lib/puppetserver/ca/action/generate.rb]: Do you mean we just shouldn't duplicate the logic? Or is there some formatting that is missing here? I think we might need it in both places because most signing will go through the API...

File: lib/puppetserver/ca/action/generate.rb (hunk 5 above)
Comment (Magisus) [reply lib/puppetserver/ca/action/generate.rb]: I think a HEREDOC would be more  idiomatic. I'll update.

File: lib/puppetserver/ca/action/generate.rb (hunk 4 above)
Comment (justinstoller) [reply lib/puppetserver/ca/action/generate.rb]: I think the format for the serial number is the same in the serial file as it is in inventory file: https://github.com/puppetlabs/puppetserver/blob/master/src/clj/puppetlabs/puppetserver/certificate_authority.clj#L373

File: lib/puppetserver/ca/action/generate.rb (hunk 2 above)
Comment (justinstoller) [reply lib/puppetserver/ca/action/generate.rb]: Because it's only the same as `!empty?` if the collection can contain _only_ truthy values).

When it's unambiguous that the collection can only contain truthy values (like here) then I'm fine with it.

My not-love for it comes from a concern that using it too freely will encourage us in the future to use it as a synonym for not-empty even in collections that can contain falsey values.

File: lib/puppetserver/ca/action/generate.rb (hunk 4 above)
Comment (Magisus) [reply lib/puppetserver/ca/action/generate.rb]: Ah, gotcha. I'll fix that.

File: lib/puppetserver/ca/action/generate.rb (hunk 2 above)
Comment (amysahli) [reply lib/puppetserver/ca/action/generate.rb]: ^ @justinstoller why don't you like any here?

File: spec/puppetserver/ca/action/import_spec.rb [hunk 6]
```diff
@@ -274,13 +215,9 @@
...
-        exit_code = Puppetserver::Ca::Cli.run(['import',
-                                                '--private-key', key_file,
-                                                '--cert-bundle', bundle_file,
-                                                '--crl-chain', chain_file],
-                                                stdout,
-                                                stderr)
-
+        exit_code = subject.run({ 'cert-bundle' => bundle_file,
+                                    'private-key'=> key_file,
+                                    'crl-chain' => chain_file })
```
Comment (amysahli) [inline spec/puppetserver/ca/action/import_spec.rb]: not important...jsut small spacing issue.

File: spec/puppetserver/ca/action/generate_spec.rb [hunk 7]
```diff
@@ -21,7 +19,7 @@
   subject { Puppetserver::Ca::Action::Generate.new(logger) }
 
   it 'prints the help output & returns 1 if invalid flags are given' do
-    exit_code = Puppetserver::Ca::Cli.run(['generate', '--hello'], stdout, stderr)
+    _, exit_code = subject.parse({ '--hello' => ''})
```
Comment (amysahli) [inline spec/puppetserver/ca/action/generate_spec.rb]: I notice elsewhere we use parse with an array like: ['--hello', '']

File: lib/puppetserver/ca/action/import.rb [hunk 8]
```diff
@@ -58,47 +75,36 @@ def run(input)
           FileSystem.write_file(puppet.settings[:cacrl], loader.crls, 0640)
 
           # Puppet's internal CA expects these file to exist.
-          FileSystem.ensure_file(puppet.settings[:serial], "001", 0640)
+          FileSystem.ensure_file(puppet.settings[:serial], "0x0001", 0640)
```
Comment (amysahli) [inline lib/puppetserver/ca/action/import.rb]: what's this  "0x0001" about?

Comment (Magisus) [reply lib/puppetserver/ca/action/import.rb]: The serials are supposed to be in hex, to maintain what puppet is doing now.

File: spec/puppetserver/ca/action/generate_spec.rb (hunk 7 above)
Comment (Magisus) [reply spec/puppetserver/ca/action/generate_spec.rb]: Oh right, I should change this. I remember there being a comment before about being surprised this works :P

**Comment by Magisus** [comment]
Updated.
